File: packages/control-contract/src/work-graph.ts
import type { Actor } from "./actor.js";
import type { Revision } from "./revision.js";
import type { TypedId } from "./ids.js";

/** A blocker edge, intentionally separate from parent/sub-work structure. */
export type WorkDependency = { readonly work_item_id: TypedId; readonly blocker_work_item_id: TypedId; readonly created_at: string };
/** Durable actor-attributed work communication. */
export type WorkComment = {
  readonly id: string; readonly work_item_id: TypedId; readonly actor: Actor; readonly body: string;
  readonly created_by_attempt_id: TypedId | null; readonly created_by_run_id: TypedId | null;
  readonly revision: Revision; readonly created_at: string;
};
export type WorkRelationKind = "parent" | "blocks";

File: src-tauri/crates/altai-control-protocol/src/work_graph.rs
//! Durable Work hierarchy, dependency, and comment contracts.

use crate::{Actor, AttemptId, Revision, RunId, WorkItemId};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct WorkDependency {
    pub work_item_id: WorkItemId,
    pub blocker_work_item_id: WorkItemId,
    pub created_at: String,
}

/// Durable attributed work communication. Parentage and dependencies are not
/// inferred from comments or from each other.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct WorkComment {
    pub id: String,
    pub work_item_id: WorkItemId,
    pub actor: Actor,
    pub body: String,
    pub created_by_attempt_id: Option<AttemptId>,
    pub created_by_run_id: Option<RunId>,
    pub revision: Revision,
    pub created_at: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WorkRelationKind {
    Parent,
    Blocks,
}
